utilities/transactions/write_prepared_txn_db.cc

```diff
@@ -402,24 +399,35 @@ void WritePreparedTxnDB::CheckPreparedAgainstMax(SequenceNumber new_max) {
       prepared_txns_.empty(),
       prepared_txns_.empty() ? 0 : prepared_txns_.top());
   while (!prepared_txns_.empty() && prepared_txns_.top() <= new_max) {
+    if (!locked) {
+      // Needed for pop(). Do it always before prepared_mutex_ to avoid deadlock
+      prepared_txns_.push_pop_mutex()->Lock();
+    }
+    WriteLock wl(&prepared_mutex_);
     auto to_be_popped = prepared_txns_.top();
     delayed_prepared_.insert(to_be_popped);
     ROCKS_LOG_WARN(info_log_,
                    "prepared_mutex_ overhead %" PRIu64 " (prep=%" PRIu64
                    " new_max=%" PRIu64,
                    static_cast<uint64_t>(delayed_prepared_.size()),
                    to_be_popped, new_max);
-    prepared_txns_.pop();
+    prepared_txns_.pop(true /*locked*/);
```

Comment:
Is there a potential race condition here between the `prepared_txns_.top() <= new_max` check and the pop here (ie. is it possible to store into `delayed_prepared_empty_` something greater than the new max?

Reply:
delayed_prepared_empty_ is a bool so cannot be greater than uint64_t. But if you mean prepared_txns_.top(), it is supposed to be updated after the pop otherwise the loop would continue indefinitely.

Reply:
Ah, I meant `delayed_prepared_.insert(to_be_popped);`

Reply:
Oh, yes. It is possible. Will submit a fix soon.